src/tests/idmap-api-test/idmap-api-test.c

```diff
@@ -0,0 +1,278 @@
+
+    /*
+     * Test with extreme cases using min and max values
+     * Note: There are no asserts, checks or expected values 
+     * here.  The only check is that the function doesn't return
+     * an error when called
+     */
+    OS_IdentifyObject(0x00000); 
+    OS_IdentifyObject(0xFFFFFFFF); 
+
+    /*
+     * Test Case For:
+     * int32 OS_ConvertToArrayIndex(void)
+     */
+    expected = OS_SUCCESS;
+
+    /*
+     * Check different id types and verify array indices
+     * Each Object Type index is added to an array index of its own type
+     * Each object type is checked once, and MUTEX is checked twice to 
+     * verify multiple indices
+     */
+
+    /*
+     * Test with nominal values
+     */
+    actual   = OS_ConvertToArrayIndex(task_id, &TestArrayIndex);
+    UtAssert_True(actual == expected , "OS_ConvertToArrayIndex() (%ld) == %ld ", (long)actual, (long)expected );
+    UtAssert_True(TestArrayIndex == 1 , "TestArrayIndex(%lu) == 1", (long)TestArrayIndex);
+
+    actual   = OS_ConvertToArrayIndex(queue_id, &TestArrayIndex);
+    UtAssert_True(actual == expected , "OS_ConvertToArrayIndex() (%ld) == %ld ", (long)actual, (long)expected );
+    UtAssert_True(TestArrayIndex == 1 , "TestArrayIndex(%lu) == 1", (long)TestArrayIndex);
+
+    actual   = OS_ConvertToArrayIndex(count_sem_id, &TestArrayIndex);
+    UtAssert_True(actual == expected , "OS_ConvertToArrayIndex() (%ld) == %ld ", (long)actual, (long)expected );
+    UtAssert_True(TestArrayIndex == 1 , "TestArrayIndex(%lu) == 1", (long)TestArrayIndex);
+
+    actual   = OS_ConvertToArrayIndex(bin_sem_id, &TestArrayIndex);
+    UtAssert_True(actual == expected , "OS_ConvertToArrayIndex() (%ld) == %ld ", (long)actual, (long)expected );
+    UtAssert_True(TestArrayIndex == 2 , "TestArrayIndex(%lu) == 2", (long)TestArrayIndex);
+
+    actual   = OS_ConvertToArrayIndex(mutex_id1, &TestArrayIndex);
+    UtAssert_True(actual == expected , "OS_ConvertToArrayIndex() (%ld) == %ld ", (long)actual, (long)expected );
+    UtAssert_True(TestArrayIndex == 1 , "TestArrayIndex(%lu) == 1", (long)TestArrayIndex);
+
+    actual   = OS_ConvertToArrayIndex(mutex_id2, &TestArrayIndex);
+    UtAssert_True(actual == expected , "OS_ConvertToArrayIndex() (%ld) == %ld ", (long)actual, (long)expected );
+    UtAssert_True(TestArrayIndex == 2 , "TestArrayIndex(%lu) == 2", (long)TestArrayIndex);
+
+    actual   = OS_ConvertToArrayIndex(mutex_id3, &TestArrayIndex);
+    UtAssert_True(actual == expected , "OS_ConvertToArrayIndex() (%ld) == %ld ", (long)actual, (long)expected );
+    UtAssert_True(TestArrayIndex == 3 , "TestArrayIndex(%lu) == 3", (long)TestArrayIndex);
+
+    actual   = OS_ConvertToArrayIndex(time_base_id, &TestArrayIndex);
+    UtAssert_True(actual == expected , "OS_ConvertToArrayIndex() (%ld) == %ld ", (long)actual, (long)expected );
+    UtAssert_True(TestArrayIndex == 1 , "TestArrayIndex(%lu) == 1", (long)TestArrayIndex);
```

Comment:
All of these "OS_ConvertToArrayIndex" tests should _not_ check for specific values -- there should be no assumption that the resources allocated will be in slot "1" .. they could be any value.

Rather - recommend to just check that the output is in range, that is, >=0 and <MAX (where MAX is the configured maximum number of objects for that type).

Reply:
Also, for the cases where more than one resource is allocated, the output array indices should be different from each other (but could still be any value).

Reply:
Is there something unique that could be checked to know the right one was returned?

Reply:
Not in the context of a functional test .... This API just outputs a number between 0 and MAX that doesn't overlap with any other resource of the same type currently in use.  So these would be the only two things that a functional test can verify.  To be really thorough, the test could create all the way up to MAX and ensure that all the returned indices were unique/non overlapping.

In practice, it is based on a mask of the LSBs and a MOD operation -- but that is white-box test material, specifically not the type of thing to be tested in a black-box functional test.

Reply:
Copy, concur.